File: raw_data_processing/EV_File_Export.py
# The goal of this script is to facilitate exporting integrated Sv data from Echoview from all frequencies

# Author: Delphine Mossman
# Date Created: 23 Mar 2023
# Date Last Modified: 23 June 2025

# 1. Open each EV file
# 2. Locate the Sv variables to be exported
# 3. Export each variable and save it with a filename that includes the frequency

# Required libraries
# %%
import win32com.client
import os

# %%

# Define some variables
homedir = "C:/Users/Delphine/Box/Glider Data/"
dep_name = input('Enter the full name of the deployment: ')

workdir = os.path.join(homedir,dep_name,'')

# Open connection to Echoview
EvApp = win32com.client.Dispatch("EchoviewCom.EvApplication")

# Define the year and month to work with, and where the day files are found
AZFPyear = dep_name[5:9]
basedir = os.path.join(workdir,'Echoview Files', '')

# for folder in ['202409']:
for folder in next(os.walk(basedir))[1]:
    AZFPmonth = folder[-2:]
    filedir = os.path.join(workdir,'Echoview Files', str(AZFPyear) + str(AZFPmonth))

    for file in os.listdir(filedir):
    # for file in ['ru39-20230817T1520_Echogram_12.EV']:
        if file[-2:] == 'EV':


            EvFile = EvApp.OpenFile(os.path.join(filedir,file))
            day = file[-5:-3]

            # Needs to be manually changed depending on whether the AZFP is zoop or fish configured
            varlist = ['Surface and Bottom Exclusion T1',
                       'Surface and Bottom Exclusion T2',
                       'Surface and Bottom Exclusion T3']

            freqlist = ['38','120','200']

            # varlist = ['Surface and Bottom Exclusion T1',
            #            'Surface and Bottom Exclusion T2',
            #            'Surface and Bottom Exclusion T3',
            #            'Surface and Bottom Exclusion T4']
            #
            # freqlist = ['120', '200', '455', '769']

            exportpath = os.path.join(workdir, 'Echoview CSV Export Files',
                                      str(AZFPyear) + str(AZFPmonth.zfill(2)))

            if not os.path.exists(exportpath):
                os.makedirs(exportpath)

            for var in varlist:
            # for var in ['Surface and Bottom Exclusion T4']:
                freq = freqlist[varlist.index(var)]

                exportname = 'RMI_'+str(AZFPyear) + \
                             str(AZFPmonth.zfill(2)) + \
                             '_' + day + '_' + \
                            freq + \
                            'kHz.csv'

                EvVar = EvFile.Variables.FindByName(var)

                EvVar.ExportIntegrationByCellsAll(os.path.join(exportpath,exportname))


            EvApp.CloseFile(EvFile)
        else:
            next


EvApp.Quit()

# %%

File: raw_data_processing/EV_File_Export.R
rm(list = ls())

library(tidyverse)
library(RDCOMClient)
library(R.utils)
library(tcltk)

sourceDirectory(
  "H:/dm1679/Code/R_Functions",
  modifiedOnly = F
)

## Setup
glider_dep = choose_directory(caption = "Select the glider deployment directory folder")

EV_dir = paste0(glider_dep, "/Echoview Files/")

# Needs to be manually changed depending on whether the AZFP is zoop or fish configured
# varlist = c('Surface and Bottom Exclusion T1',
#            'Surface and Bottom Exclusion T2',
#            'Surface and Bottom Exclusion T3')
# freqlist = c('38','120','200')

varlist = c('Surface and Bottom Exclusion T1',
           'Surface and Bottom Exclusion T2',
           'Surface and Bottom Exclusion T3',
           'Surface and Bottom Exclusion T4')
freqlist = c('120', '200', '455', '769')

EVAppObj = COMCreate('EchoviewCom.EvApplication')

for(folder in list.files(EV_dir, recursive = F, 
                         pattern = "[0-9]{6}", full.names = T, 
                         include.dirs = T)) {
  for(file in list.files(folder, full.names = T, pattern = "*.EV")) {
    # Open each file
    EVFile = EVAppObj$OpenFile(file)
    
    # Get the day of the month from the file name
    day = substr(file, nchar(file)-4, nchar(file)-3)
    
    # Create the export directory, if it does not already exist
    create_dir(gsub("Echoview Files", "Echoview CSV Export Files", folder))
    
    for(variable in varlist){
      # Index-match the frequency to the variable
      freq = freqlist[which(varlist == variable)]
      
      export_file_name = paste0(gsub("Echoview Files", "Echoview CSV Export Files", folder),
                                "/RMI_",substr(folder, nchar(folder)-5, nchar(folder)),
                                "_",day,"_",freq,"kHz.csv")
      
      EVVar = EVFile[["Variables"]]$FindByName(variable)
      
      EVVar$ExportIntegrationByCellsAll(export_file_name)
      
    }
    EVFile$Close()
  }
}

EVAppObj$Quit()
